### PR TITLE
SQL compatible object fetch api

### PR DESCRIPTION
**Pull request by snopoke**
@czue @benrudolph 
buddy @emord

:fish:  / :post_office:

### CODE REVIEW

File: corehq/apps/api/object_fetch_api.py [hunk 1]
```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
...
+                            "img": '1',
+                            "size": fsize,
+                            "max_size": max_filesize,
+                            "max_image_width": max_width,
+                            "max_image_height": max_height
+                        })
+                        r.write('<img src="%(attach_url)s?%(params)s">' % {
+                                "attach_url": url_base,
+                                "params": url_params
+                        })
```
Comment (dimagimon) [inline corehq/apps/api/object_fetch_api.py]: E124 closing bracket does not match visual indentation

File: corehq/apps/api/object_fetch_api.py [hunk 2]
```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
...
+        except AttachmentNotFound as e:
+            print e
             raise Http404
         
-        headers = resp.resp.headers
-        content_type = headers.get('Content-Type', None)
+        return StreamingHttpResponse(streaming_content=content.content_stream, content_type=content.content_type)
+
+
+def fetch_case_image(domain, case_id, attachment_id, filesize_limit=0, width_limit=0, height_limit=0, fixed_size=None):
```
Comment (dimagimon) [inline corehq/apps/api/object_fetch_api.py]: E501 line too long (119 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark: though this could probably use a test or 2

File: corehq/apps/api/object_fetch_api.py [hunk 3]
```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
...
+
+        return StreamingHttpResponse(streaming_content=attachment_stream, content_type=mime_type)
 
 
 class FormAttachmentAPI(View):
     @method_decorator(login_or_digest_or_basic_or_apikey())
-    def get(self, *args, **kwargs):
-        form_id = kwargs.get('form_id', None)
-        if not form_id or not XFormInstance.get_db().doc_exist(form_id):
+    def get(self, request, domain, form_id=None, attachment_id=None):
```
Comment (benrudolph) [inline corehq/apps/api/object_fetch_api.py]: +1 for real args

File: corehq/apps/api/object_fetch_api.py [hunk 4]
```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
...
             raise Http404
 
-        attachment_key = kwargs.get('attachment_id', None)
-
         try:
-            resp = XFormInstance.get_db().fetch_attachment(form_id, attachment_key, stream=True)
-        except ResourceNotFound:
+            content = FormAccessors(domain).get_attachment_content(form_id, attachment_id)
+        except AttachmentNotFound as e:
+            print e
```
Comment (benrudolph) [inline corehq/apps/api/object_fetch_api.py]: print

File: corehq/apps/api/object_fetch_api.py [hunk 5]
```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
...
+    meta, stream = cached_image.get(size_key=size_key)
+
+    if do_constrain:
+        def meets_constraint(constraints, meta):
+            for c, limit in constraints.items():
+                if meta[c] > limit:
+                    return False
+            return True
+
+        if meets_constraint(constraint_dict, meta):
```
Comment (benrudolph) [inline corehq/apps/api/object_fetch_api.py]: this seems a little odd, maybe just have `if not meets_constraint(constraint_dict, meta):`

File: corehq/form_processor/backends/couch/dbaccessors.py [hunk 6]
```diff
@@ -113,3 +134,18 @@ def get_last_modified_dates(domain, case_ids):
...
+        return _get_attachment_content(CommCareCase, case_id, attachment_id)
+
+
+def _get_attachment_content(doc_class, doc_id, attachment_id):
+    try:
+        resp = doc_class.get_db().fetch_attachment(doc_id, attachment_id, stream=True)
+    except ResourceNotFound:
+        raise AttachmentNotFound(attachment_id)
+
+    headers = resp.resp.headers
```
Comment (benrudolph) [inline corehq/form_processor/backends/couch/dbaccessors.py]: just making sure this isn't a typo

**Comment by benrudolph** [comment]
a few minor comments. agree a couple tests would be nice. gonna merge because most of it is not in use

File: corehq/form_processor/backends/couch/dbaccessors.py (hunk 6 above)
Comment (snopoke) [inline corehq/form_processor/backends/couch/dbaccessors.py]: nope, not a typo :)

**Comment by snopoke** [comment]
PR feedback https://github.com/dimagi/commcare-hq/pull/10353

**Pull request merged**